### DISCUSSION TITLE
Index 2 is out of bounds for dimension 1 with size 2

### DISCUSSION THREAD

**Discussion by Applewine33** [Project-MONAI/MONAI#7494] [General]
I try to run this code while when it run to "loss = loss_function(outputs, labels)" will get error and print this RuntimeError "Index 2 is out of bounds for dimension 1 with size 2". How can I fix it?

max_epochs = 100
val_interval = 2
best_metric = -1
best_metric_epoch = -1
epoch_loss_values = []
metric_values = []
post_pred = Compose([AsDiscrete(argmax=True, to_onehot=2)])
post_label = Compose([AsDiscrete(to_onehot=2)])

for epoch in range(max_epochs):
    print("-" * 10)
    print(f"epoch {epoch + 1}/{max_epochs}")
    model.train()
    epoch_loss = 0
    step = 0
    for batch_data in train_loader:
        step += 1
        inputs, labels = (
            batch_data["image"].to(device),
            batch_data["label"].to(device),
        )
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = loss_function(outputs, labels)
        loss.backward()
        optimizer.step()
        epoch_loss += loss.item()
        print(f"{step}/{len(train_ds) // train_loader.batch_size}, " f"train_loss: {loss.item():.4f}")
    epoch_loss /= step
    epoch_loss_values.append(epoch_loss)
    print(f"epoch {epoch + 1} average loss: {epoch_loss:.4f}")

    if (epoch + 1) % val_interval == 0:
        model.eval()
        with torch.no_grad():
            for val_data in val_loader:
                val_inputs, val_labels = (
                    val_data["image"].to(device),
                    val_data["label"].to(device),
                )
                roi_size = (160, 160, 160)
                sw_batch_size = 4
                val_outputs = sliding_window_inference(val_inputs, roi_size, sw_batch_size, model)
                val_outputs = [post_pred(i) for i in decollate_batch(val_outputs)]
                val_labels = [post_label(i) for i in decollate_batch(val_labels)]
                # compute metric for current iteration
                dice_metric(y_pred=val_outputs, y=val_labels)

            # aggregate the final mean dice result
            metric = dice_metric.aggregate().item()
            # reset the status for next validation round
            dice_metric.reset()

            metric_values.append(metric)
            if metric > best_metric:
                best_metric = metric
                best_metric_epoch = epoch + 1
                torch.save(model.state_dict(), os.path.join(images_dir, "best_metric_model.pth"))
                print("saved new best metric model")
            print(
                f"current epoch: {epoch + 1} current mean dice: {metric:.4f}"
                f"\nbest mean dice: {best_metric:.4f} "
                f"at epoch: {best_metric_epoch}"
            )

**Comment by KumoLiu** [comment]
Hi @Applewine33, thanks for your interest here.
The error you're encountering, RuntimeError: Index 2 is out of bounds for dimension 1 with size 2, generally happens when the code tries to access index 2 of dimension 1, but that dimension only has a size of 2 (meaning it only has valid indices 0 and 1).
In your specific case of this occurring at the line loss = loss_function(outputs, labels), this indicates that there most likely is a mismatch between the number of classes in your model output and the number of classes you have in your target labels.
Here are a few things you can check:
Output Classes: Ensure that the number of output units in the last layer of your network matches the number of classes in your dataset.
Label Values: Check your label tensor to make sure that it does not contain any values higher than (number_of_output_units - 1). If your model only outputs 2 classes (indices 0 and 1), you should check that the labels tensor does not contain any values higher than 1.

Thanks.
(more like a usage issue, move it to discussion)

**Reply by Applewine33** [reply]
I find what wrong and fix it. But while I count train loss, it shows that "The size of the proposed random crop ROI is larger than the image size". I try to change the spatial_size in RandCropByPosNegLabeld to be smaller but it do not work. How should I do?

**Reply by KumoLiu** [reply]
Ensure the ROI size smaller than the image size. Or You can set `allow_smaller=True` to skip the error.
https://github.com/Project-MONAI/MONAI/blob/f6f9e817892090f6c1647f60f51fbc5ed80b74f1/monai/transforms/croppad/dictionary.py#L919

**Reply by Applewine33** [reply]
I try this already. But still have the same error.
